src/main/java/dan200/computercraft/shared/network/client/TerminalState.java

```diff
@@ -0,0 +1,180 @@
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+
+/**
+ * A snapshot of a terminal's state.
+ *
+ * This is somewhat memory inefficient (we build a buffer, only to write it elsewhere), however it means we get a
+ * complete and accurate description of a terminal, which avoids a lot of complexities with resizing terminals, dirty
+ * states, etc...
+ */
+public class TerminalState
+{
+    public final boolean colour;
+
+    public final int width;
+    public final int height;
+
+    private final boolean compress;
+
+    @Nullable
+    private final ByteBuf buffer;
+
+    private ByteBuf compressed;
+
+    public TerminalState( boolean colour, @Nullable Terminal terminal )
+    {
+        this( colour, terminal, true );
+    }
+
+    public TerminalState( boolean colour, @Nullable Terminal terminal, boolean compress )
+    {
+        this.colour = colour;
+        this.compress = compress;
+
+        if( terminal == null )
+        {
+            this.width = this.height = 0;
+            this.buffer = null;
+        }
+        else
+        {
+            this.width = terminal.getWidth();
+            this.height = terminal.getHeight();
+
+            ByteBuf buf = this.buffer = Unpooled.buffer();
+            terminal.write( new PacketBuffer( buf ) );
+        }
+    }
+
+    public TerminalState( PacketBuffer buf )
+    {
+        this.colour = buf.readBoolean();
+        this.compress = buf.readBoolean();
+
+        if( buf.readBoolean() )
+        {
+            this.width = buf.readVarInt();
+            this.height = buf.readVarInt();
+
+            int length = buf.readVarInt();
+            this.buffer = readCompressed( buf, length, compress );
+        }
+        else
+        {
+            this.width = this.height = 0;
+            this.buffer = null;
+        }
+    }
+
+    public void write( PacketBuffer buf )
+    {
+        buf.writeBoolean( colour );
+        buf.writeBoolean( compress );
+
+        buf.writeBoolean( buffer != null );
+        if( buffer != null )
+        {
+            buf.writeVarInt( width );
+            buf.writeVarInt( height );
+
+            ByteBuf sendBuffer = getCompressed();
+            int index = sendBuffer.readerIndex();
+            buf.writeVarInt( sendBuffer.readableBytes() );
+            buf.writeBytes( sendBuffer );
+            sendBuffer.readerIndex( index );
+        }
+    }
+
+    public boolean hasTerminal()
+    {
+        return buffer != null;
+    }
+
+    public void apply( Terminal terminal )
+    {
+        if( buffer == null ) throw new NullPointerException( "buffer" );
+        terminal.read( new PacketBuffer( buffer ) );
+    }
+
+    private ByteBuf getCompressed()
+    {
+        if( buffer == null ) throw new NullPointerException( "buffer" );
+        if( !compress ) return buffer;
+        if( compressed != null ) return compressed;
+
+        ByteBuf compressed = this.compressed = Unpooled.directBuffer();
```

Comment:
I do not this this is safe. Packets may be sent on multiple threads at once, and so this is could lead to a race condition with multiple threads reading from `buffer`.

I kind of want to avoid compressing on the main thread (and compressing if there's no one watching this chunk), but that may be a premature optimisation.